crates/shared/src/price_estimation/native_price_cache.rs

```diff
@@ -0,0 +1,317 @@
+use crate::price_estimation::PriceEstimationError;
+use primitive_types::H160;
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock, Weak};
+use std::time::{Duration, Instant};
+
+#[derive(Debug, Clone)]
+struct CachedPrice {
+    price: f64,
+    updated_at: Instant,
+}
+
+struct Inner {
+    cache: RwLock<HashMap<H160, CachedPrice>>,
+    estimator: Box<dyn NativePriceEstimating>,
+    max_age: Duration,
+}
+
+impl Inner {
+    async fn estimate_prices_and_update_cache(
+        &self,
+        tokens: &[H160],
+    ) -> Vec<Result<f64, PriceEstimationError>> {
+        if tokens.is_empty() {
+            return Vec::default();
+        }
+
+        let now = Instant::now();
+        let prices = self.estimator.estimate_native_prices(tokens).await;
+        {
+            let mut cache = self.cache.write().unwrap();
+            for (token, result) in tokens.iter().zip(prices.iter()) {
+                if let Ok(price) = result {
+                    let mut entry = cache.entry(*token).or_insert_with(|| CachedPrice {
+                        price: *price,
+                        updated_at: now,
+                    });
+                    entry.updated_at = now;
+                    entry.price = *price;
+                }
+            }
+        }
+        prices
+    }
+
+    fn get_cached_prices(&self, tokens: &[H160]) -> Vec<Option<f64>> {
+        if tokens.is_empty() {
+            return Vec::default();
+        }
+
+        let now = Instant::now();
+        let cache = self.cache.read().unwrap();
+        tokens
+            .iter()
+            .map(|token| match cache.get(token) {
+                Some(entry) if now.saturating_duration_since(entry.updated_at) < self.max_age => {
+                    Some(entry.price)
+                }
+                _ => None,
+            })
+            .collect()
+    }
+}
+
+/// Wrapper around `Box<dyn PriceEstimating>` which caches successful price estimates for some time
+/// and supports updating the cache in the background.
+/// The size of the underlying cache is unbounded.
+pub struct CachingNativePriceEstimator(Arc<Inner>);
+
+impl CachingNativePriceEstimator {
+    /// Creates new CachingNativePriceEstimator using `estimator` to calculate native prices which
+    /// get cached a duration of `max_age`.
+    pub fn new(estimator: Box<dyn NativePriceEstimating>, max_age: Duration) -> Self {
+        Self(Arc::new(Inner {
+            estimator,
+            cache: RwLock::new(Default::default()),
+            max_age,
+        }))
+    }
+
+    /// Spawns a background task maintaining the cache once per `update_interval`.
+    /// Only outdated prices get updated and older prices have a higher priority.
+    /// If `update_size` is `Some(n)` at most `n` prices get updated per interval.
+    /// If `update_size` is `None` no limit gets applied.
+    pub fn with_maintenance(self, update_interval: Duration, update_size: Option<usize>) -> Self {
```

Comment:
I think this doesn't have to consume `self`. Could take by normal reference.

Reply:
No, it doesn't **have** to consume `self`. I just thought it would be nicer to read if you are able to chain those calls when you create a new estimator.
```rust
let estimator = CachingNativePriceEstimator::new(...).with_maintenenace(...);
```
vs
```rust
let estimator = CachingNativePriceEstimator::new(...);
estimator.with_maintenance(...);
```
If that is not desired I would prefer the function to be renamed as well:
```rust
pub fn spawn_maintenance_task(&self, update_interval: Duration, update_size: Option<usize>)
```

Reply:
Unless you explicitly want to use the builder pattern I prefer the third version. Usually you would use it when there are lots of things to set.
The difference is small so either way is fine I just slightly prefer it that way but we can continue with current version too.

Reply:
I don't really care that much but I wanted to explain my reasoning.
Changed the signature and name now.